scripts/servers.py

```diff
@@ -0,0 +1,593 @@
+from scripts import common
+
+
+@contextlib.contextmanager
+def managed_process(
+        command_args, human_readable_name='Process', shell=False,
+        timeout_secs=60, **popen_kwargs):
+    """Context manager for starting and stopping a process gracefully.
+
+    Args:
+        command_args: list(int|str). A sequence of program arguments, where the
+            program to execute is the first item. Ints are allowed in order to
+            accomodate e.g. port numbers.
+        human_readable_name: str. The human-readable name of the process. Used
+            by the function's logging logic to improve readability.
+        shell: bool. Whether the command should be run inside of its own shell.
+            WARNING: Executing shell commands that incorporate unsanitized input
+            from an untrusted source makes a program vulnerable to
+            [shell injection](https://w.wiki/_Ac2), a serious security flaw
+            which can result in arbitrary command execution. For this reason,
+            the use of `shell=True` is **strongly discouraged** in cases where
+            the command string is constructed from external input.
+        timeout_secs: int. The time allotted for the managed process and its
+            descendants to terminate themselves. After the timeout, any
+            remaining processes will be killed abruptly.
+        **popen_kwargs: dict(str: *). Same kwargs as `subprocess.Popen`.
+
+    Yields:
+        psutil.Process. The process managed by the context manager.
+    """
+    # TODO(#11549): Move this to top of the file.
+    if common.PSUTIL_DIR not in sys.path:
+        sys.path.insert(1, common.PSUTIL_DIR)
+    import psutil
+
+    get_debug_info = lambda p: (
+        '%s(name="%s", pid=%d)' % (human_readable_name, p.name(), p.pid)
+        if p.is_running() else '%s(pid=%d)' % (human_readable_name, p.pid,))
+
+    stripped_args = (('%s' % arg).strip() for arg in command_args)
+    non_empty_args = (s for s in stripped_args if s)
+
+    command = ' '.join(non_empty_args) if shell else list(non_empty_args)
+    python_utils.PRINT('Starting new %s: %s' % (human_readable_name, command))
+    popen_proc = psutil.Popen(command, shell=shell, **popen_kwargs)
+
+    try:
+        yield popen_proc
+    finally:
+        python_utils.PRINT('Stopping %s...' % get_debug_info(popen_proc))
+        procs_still_alive = [popen_proc]
+        try:
+            if popen_proc.is_running():
+                # Children must be terminated before the parent, otherwise they
+                # may become zombie processes.
+                procs_still_alive = (
+                    popen_proc.children(recursive=True) + [popen_proc])
+
+            procs_to_kill = []
+            for proc in procs_still_alive:
+                if proc.is_running():
+                    logging.info('Terminating %s...' % get_debug_info(proc))
+                    proc.terminate()
+                    procs_to_kill.append(proc)
```

Comment:
@brianrodri I think there's a race condition here because the process could terminate after the `if` but before we call `proc.terminate()`. Then we'll get a `NoSuchProcess` exception

Reply:
That's true, but there's no real way to avoid it unfortunately. If such an error occurs, then the outer try/catch will handle it and report the error.

Reply:
nvm. I missed the `try...except` wrapping all this code